app/utils/log_parser.py

```diff
@@ -530,82 +530,72 @@ def _clean_path(line):
-            err_msg = "Cannot read build log file for %s-%s-%s"
-            utils.LOG.exception(ex)
-            utils.LOG.error(err_msg, job, kernel, defconfig)
-            status = 500
-            ERR_ADD(errors, status, err_msg % (job, kernel, defconfig))
-        else:
-            try:
-                # TODO: count the lines here.
-                if error_lines:
-                    with io.open(errors_file, mode="w") as w_file:
-                        for line in error_lines:
-                            w_file.write(line)
-                            w_file.write(u"\n")
-                if warning_lines:
-                    with io.open(warnings_file, mode="w") as w_file:
-                        for line in warning_lines:
-                            w_file.write(line)
-                            w_file.write(u"\n")
-                if mismatch_lines:
-                    with io.open(mismatches_file, mode="w") as w_file:
-                        for line in mismatch_lines:
-                            w_file.write(line)
-                            w_file.write(u"\n")
-            except IOError, ex:
-                err_msg = "Error writing to errors/warnings file for %s-%s-%s"
-                utils.LOG.exception(ex)
-                utils.LOG.error(err_msg, job, kernel, defconfig)
-                status = 500
-                ERR_ADD(errors, status, err_msg % (job, kernel, defconfig))
-    else:
-        utils.LOG.warn("Build dir '%s' does not have a build log" % defconfig)
+    utils.LOG.info("Parsing build log file '%s'", log_file)
+
+    err_append = error_lines.append
+    warn_append = warning_lines.append
+    mismatch_append = mismatch_lines.append
+
+    try:
+        with io.open(log_file) as read_file:
+            for line in read_file:
+                if any(re.search(err_pattrn, line)
+                       for err_pattrn in ERROR_PATTERNS):
+                    line = line.strip()
+                    err_append(_clean_path(line))
+                    continue
+
+                if re.search(WARNING_PATTERN, line) \
+                        and not any(re.search(warn_pattrn, line)
+                                    for warn_pattrn in EXCLUDE_PATTERNS):
+                    line = line.strip()
+                    warn_append(_clean_path(line))
+                    continue
+
+                if re.search(MISMATCH_PATTERN, line):
+                    line = line.strip()
+                    mismatch_append(_clean_path(line))
+    except IOError, ex:
+        err_msg = "Cannot read build log file for %s-%s-%s"
+        utils.LOG.exception(ex)
+        utils.LOG.error(err_msg, job, kernel, defconfig)
         status = 500
-
+        ERR_ADD(errors, status, err_msg % (job, kernel, defconfig))
+        return status, [], [], []
+
+    def _save_lines(lines, filename):
+        # TODO: count the lines here.
+        if not lines:
+            return
+        with open(filename, mode="w") as w_file:
```

Comment:
I switched from io.open() to builtin open(), as the latter is immediately recognizable to anyone familiar with Python. I didn't encounter any other project which would explicitly use io.open(), so I assumed using the clearer open() is safe here.